### PR TITLE
Add Common Utilities Package

### PR DESCRIPTION
**Pull request by amery**
### **User description**
# Add Common Utilities Package

## Summary

This PR establishes the common package with essential shared utilities for the
protomcp.org ecosystem, including memory-safe slice operations and an
error-free string building utility.

### Core Utilities

- **Slice Utilities**: Generic functions for preventing memory leaks when
  truncating slices containing pointers or reference types.
  - `ClearSlice`: Zeros all elements including unused capacity, returns empty
    slice that reuses the underlying array.
  - `ClearAndNilSlice`: Zeros all elements and returns nil, releasing the
    underlying array for garbage collection.

- **String Building**: LazyBuffer provides a convenient wrapper around
  strings.Builder that eliminates error handling.
  - Chain-able methods for incremental string building.
  - Nil-safe operations.
  - Optimised for code generation and template rendering.

### Infrastructure

- **Test Migration**: Replaced local testutils package with darvaza.org/core
  utilities. 
  - Implemented MANDATORY TestCase interface pattern for data-driven tests.
  - Achieved 100% test coverage across all utilities.

- **Build System**: Imported latest build system improvements from
  darvaza.org/slog.
  - Updated make targets and CI configuration.
  - Synchronised linting and formatting rules.

## Testing

All utilities have comprehensive test coverage including:

- Table-driven tests using darvaza.org/core TestCase patterns.
- Edge cases for nil, empty, and large inputs.
- Memory leak prevention validation.
- Performance benchmarks comparing against standard library alternatives.

## Compatibility

- Requires Go 1.23 or later.
- Uses darvaza.org/core v0.18.1 for test infrastructure.
- No breaking changes to existing APIs.

___

### **PR Type**
Enhancement

___

### **Description**
- Add slice utilities for memory leak prevention

- Add LazyBuffer for error-free string building

- Update build system with coverage improvements

- Remove empty testutils package

___

### Diagram Walkthrough

```mermaid
flowchart LR
  A["Remove testutils"] --> B["Add slice utilities"]
  B --> C["Add LazyBuffer"]
  C --> D["Update build system"]
  D --> E["Enhanced coverage"]
```

<details> <summary><h3> File Walkthrough</h3></summary>

<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Documentation</strong></td><td><details><summary>4 files</summary><table>
<tr>
  <td><strong>doc.go</strong><dd><code>Expand package documentation with utilities overview</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/protomcp/common/pull/1/files#diff-a20b1b3b4b2bca5bef2b853a6e3f19def513381f9c7ee68d6979f0435c885dcf">+50/-0</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>README.md</strong><dd><code>Add comprehensive documentation for new utilities</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/protomcp/common/pull/1/files#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5">+181/-6</a>&nbsp; </td>

</tr>

<tr>
  <td><strong>AGENT.md</strong><dd><code>Update testing practices documentation</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/protomcp/common/pull/1/files#diff-c1e87407ce72f38b645f627c7fbda6e6999170c840cded6ce82c1bd2c10923e8">+2/-2</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>README-coverage.md</strong><dd><code>Update coverage documentation</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/protomcp/common/pull/1/files#diff-806bed75a9e0b7ba450ee7ca3edbdbd7b2bd70087768424c67a2f94ad450a63a">+14/-19</a>&nbsp; </td>

</tr>
</table></details></td></tr><tr><td><strong>Enhancement</strong></td><td><details><summary>11 files</summary><table>
<tr>
  <td><strong>slices.go</strong><dd><code>Add ClearSlice and ClearAndNilSlice functions</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/protomcp/common/pull/1/files#diff-2669a4758be94eb1255e2b6631c49b098d86444cce8f276df63434916aa98663">+43/-0</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>lazybuffer.go</strong><dd><code>Add LazyBuffer wrapper for strings.Builder</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/protomcp/common/pull/1/files#diff-d0cfde05cbe445e5d82e51e77d4587d2573a4cb65d82efc200e4e6f5367a44d6">+87/-0</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>make_coverage.sh</strong><dd><code>Rewrite coverage script for single module execution</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/protomcp/common/pull/1/files#diff-f28b7742d2098e7eb569772726262d78e0be2928903ac9191e86134845f3a189">+98/-83</a>&nbsp; </td>

</tr>

<tr>
  <td><strong>merge_coverage.sh</strong><dd><code>Add script to merge coverage profiles</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/protomcp/common/pull/1/files#diff-9310ef500b186d4fc506634d757075f5b5c99d9b40087ac1e2689341769cf828">+35/-0</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>make_codecov.sh</strong><dd><code>Simplify codecov script generation</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/protomcp/common/pull/1/files#diff-2c93095d48445ab66e1dba9348ed6de05f4ee51edfe094b590f6e3b48268a8d9">+19/-99</a>&nbsp; </td>

</tr>

<tr>
  <td><strong>gen_mk.sh</strong><dd><code>Add coverage and race targets to makefile</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/protomcp/common/pull/1/files#diff-32ce7f2873afedd7004764ebb6e2fefcbb5bda83829b97e77edac1eb0007fb0c">+25/-2</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>Makefile</strong><dd><code>Update coverage targets and clean-coverage rule</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/protomcp/common/pull/1/files#diff-76ed074a9305c04054cdebb9e9aad2d818052b07091de1f20cad0bbac34ffb52">+27/-32</a>&nbsp; </td>

</tr>

<tr>
  <td><strong>codecov.yml</strong><dd><code>Add Go module caching and clean coverage</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/protomcp/common/pull/1/files#diff-512f30627f95d7fcdc76dc362d896351e9a4539994b7ab00b41979e653a6d0a9">+3/-2</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>race.yml</strong><dd><code>Add dedicated race detection workflow</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/protomcp/common/pull/1/files#diff-a1501cd3f8cf73008680e5f2725d3c48d0501da0df80d12faad162cc72e517ab">+25/-0</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>test.yml</strong><dd><code>Simplify test workflow and add caching</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/protomcp/common/pull/1/files#diff-faff1af3d8ff408964a57b2e475f69a6b7c7b71c9978cccc8f471798caac2c88">+3/-16</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>build.yml</strong><dd><code>Add Go module caching to build workflow</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/protomcp/common/pull/1/files#diff-5c3fa597431eda03ac3339ae6bf7f05e1a50d6fc7333679ec38e21b337cb6721">+2/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>
</table></details></td></tr><tr><td><strong>Tests</strong></td><td><details><summary>2 files</summary><table>
<tr>
  <td><strong>slices_test.go</strong><dd><code>Comprehensive tests for slice utilities</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/protomcp/common/pull/1/files#diff-adea7af782c095a24f7e77fa73291eaa35d7078f4830b173728f34e1b8c8847a">+424/-0</a>&nbsp; </td>

</tr>

<tr>
  <td><strong>lazybuffer_test.go</strong><dd><code>Complete test suite for LazyBuffer</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/protomcp/common/pull/1/files#diff-1572d389d9459b7349c9f5179b7e6024279bb4d6e3cb8bf6e84322dffc2b5ade">+424/-0</a>&nbsp; </td>

</tr>
</table></details></td></tr><tr><td><strong>Miscellaneous</strong></td><td><details><summary>1 files</summary><table>
<tr>
  <td><strong>doc.go</strong><dd><code>Remove empty testutils package</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/protomcp/common/pull/1/files#diff-332b86d465498a310091744b2b4c0c3f86a4c1617e4e67fea1b7707f6e7e5a74">+0/-2</a>&nbsp; &nbsp; &nbsp; </td>

</tr>
</table></details></td></tr><tr><td><strong>Dependencies</strong></td><td><details><summary>2 files</summary><table>
<tr>
  <td><strong>go.mod</strong><dd><code>Add darvaza.org/core dependency for testing</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/protomcp/common/pull/1/files#diff-33ef32bf6c23acb95f5902d7097b7a1d5128ca061167ec0716715b0b9eeaa5f6">+7/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>go.sum</strong><dd><code>Update dependency checksums</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/protomcp/common/pull/1/files#diff-3295df7234525439d778f1b282d146a4f1ff6b415248aaac074e8042d9f42d63">+6/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>
</table></details></td></tr><tr><td><strong>Configuration changes</strong></td><td><details><summary>4 files</summary><table>
<tr>
  <td><strong>gen_index.sh</strong><dd><code>Update grouping configuration</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/protomcp/common/pull/1/files#diff-94bb71a4a33e98bffab7188f2ff73323f4fdb66daa817a9efaa9dc1251c08d52">+2/-2</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>.editorconfig</strong><dd><code>Modernize EditorConfig settings</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/protomcp/common/pull/1/files#diff-0947e2727d6bad8cd0ac4122f5314bb5b04e337393075bc4b5ef143b17fcbd5b">+10/-15</a>&nbsp; </td>

</tr>

<tr>
  <td><strong>.golangci.yml</strong><dd><code>Update golangci-lint configuration syntax</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/protomcp/common/pull/1/files#diff-6179837f7df53a6f05c522b6b7bb566d484d5465d9894fb04910dd08bb40dcc9">+2/-3</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>cspell.json</strong><dd><code>Update spell check dictionary</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/protomcp/common/pull/1/files#diff-3094120fcb2e87cf154fce1053bc486a13eaa9d4200a24f858bc9c39ec2e0936">+4/-2</a>&nbsp; &nbsp; &nbsp; </td>

</tr>
</table></details></td></tr></tr></tbody></table>

</details>

___

## Summary by CodeRabbit

* **New Features**
  * Added generic slice utility functions to clear and release slices, helping prevent memory leaks.
  * Introduced `LazyBuffer`, a wrapper for simplified and efficient string building without error handling.
  * Added a new GitHub Actions workflow for race detection.
  * Enabled caching in Go setup steps across CI workflows to speed up dependency resolution.

* **Documentation**
  * Expanded and clarified documentation for slice utilities and `LazyBuffer` with detailed usage examples.
  * Updated README to reflect new features and improved integration guidance.
  * Improved and generalized coverage testing documentation.

* **Bug Fixes**
  * Enhanced handling and merging of Go coverage profiles, improving reliability in CI workflows.
  * Improved robustness of Codecov upload scripts with better error handling and simplified configuration.

* **Refactor**
  * Consolidated and reorganized Makefile variables and targets for better maintainability.
  * Simplified and streamlined internal build scripts for coverage and race detection.
  * Restructured linter configuration for clarity and updated to latest version.

* **Chores**
  * Updated dependencies and configuration files for consistency and improved caching in CI.
  * Improved spell-check and code style configurations.
  * Removed local test utilities references in favor of external package integration.

* **Tests**
  * Introduced comprehensive unit tests and benchmarks for new slice and string utilities.
  * Enhanced CI workflows with dedicated race detection and improved test job strategies.

### CODE REVIEW

File: .github/workflows/codecov.yml [hunk 1]
```diff
@@ -19,9 +19,10 @@ jobs:
       with:
         go-version: '1.23'
         cache: true
+        cache-dependency-path: '**/go.sum'
 
     - name: Run tests with coverage
-      run: make codecov GOTEST_FLAGS="-covermode=atomic"
+      run: make clean-coverage codecov
```
[the remark this answers is not in the document]
Comment (amery) [reply .github/workflows/codecov.yml:25]: can the codecov/codecov-action@v4 action truly replace our generated codecov.sh?

File: .golangci.yml [hunk 2]
```diff
@@ -1,11 +1,10 @@
 run:
-  timeout: 5m
+  deadline: 5m
   allow-parallel-runners: true
 
 linters-settings:
   govet:
-    enable:
-      - fieldalignment
+    enable=fieldalignment: true
```
[the remark this answers is not in the document]
Comment (amery) [reply .golangci.yml]: check now

File: internal/build/make_coverage.sh [hunk 3]
```diff
@@ -1,107 +1,122 @@
 #!/bin/sh
-# shellcheck disable=SC1007,SC3043 # empty assignments, -o and local usage
+# shellcheck disable=SC1007,SC3043 # empty assignments and local usage
```
[the remark this answers is not in the document]
Comment (amery) [reply internal/build/make_coverage.sh:2]: non-POSIX but works on EVERY sh implementation
